tools/ci-build/publisher/src/sort.rs

```diff
@@ -34,27 +34,28 @@ pub fn dependency_order(packages: Vec<Package>) -> Result<Vec<Package>> {
 
     Ok(order
         .into_iter()
-        .map(&mut |handle| packages.remove(&handle).unwrap())
+        .map(&mut |handle: PackageHandle| packages.remove(&handle.name).unwrap())
         .collect())
 }
 
 fn dependency_order_visit(
     package_handle: &PackageHandle,
-    packages: &BTreeMap<PackageHandle, Package>,
-    stack: &mut BTreeSet<PackageHandle>,
-    visited: &mut BTreeSet<PackageHandle>,
+    packages: &BTreeMap<String, Package>,
+    stack: &mut BTreeSet<String>,
+    visited: &mut BTreeSet<String>,
```

Comment:
Does this break anything if we depend on multiple versions of the same third-party crate?

Reply:
I don't think it does because the `publisher` is currently (and I think will continue to be) concerned with crates that we are allowed to publish. Whenever `publisher` discovers & validates packages, it deals with [such packages](https://github.com/awslabs/smithy-rs/blob/d800d33e283ff10d9d75ec81f933a5137ef14721/tools/ci-build/publisher/src/package.rs#L151).

So the third-party crates have not been in the picture for discovery & validation of packages. For those we can publish, multiple version checks have already been done by [validate_packages](https://github.com/awslabs/smithy-rs/blob/d800d33e283ff10d9d75ec81f933a5137ef14721/tools/ci-build/publisher/src/package.rs#L153C17-L153C17) before we determine the sort order.

Reply:
I don't think it does because the `publisher` is currently (and I think will continue to be) concerned with crates that we are allowed to publish. Whenever `publisher` discovers & validates packages, it deals with [such packages](https://github.com/awslabs/smithy-rs/blob/d800d33e283ff10d9d75ec81f933a5137ef14721/tools/ci-build/publisher/src/package.rs#L151).

So the third-party crates have not been in the picture for discovery & validation of packages. For those we can publish, multiple version checks have already been done by [validate_packages](https://github.com/awslabs/smithy-rs/blob/d800d33e283ff10d9d75ec81f933a5137ef14721/tools/ci-build/publisher/src/package.rs#L153C17-L153C17) before we determine the sort order.

Should the requirements for `publish` change in the future to handle third-party crate as well, an enhancement we can make is to define custom `Eq` and `PartialEq` traits for `crate::package::Version`.